src/XrdCeph/XrdCephBulkAioRead.cc

```diff
@@ -0,0 +1,237 @@
+   *
+   * @return  zero on success, negative error code on failure
+   *
+   */
+
+  std::string obj_name;
+  size_t obj_idx;
+  librados::AioCompletion* cmpl;
+  librados::ObjectReadOperation* op;
+
+
+  for (auto const& op_data: operations) {
+    obj_idx = op_data.first;
+    //16 bytes for object hex number, 1 for dot and 1 for null-terminator
+    char object_suffix[18];
+    int sp_bytes_written;
+    sp_bytes_written = snprintf(object_suffix, sizeof(object_suffix), ".%016zx", obj_idx);
+    if (sp_bytes_written > (int) sizeof(object_suffix)) {
+      log_func((char*)"Can not fit object suffix into buffer for file %s -- too big\n", file_name.c_str());
+      return -EFBIG;
+    } 
+
+    try {
+      obj_name =  file_name + std::string(object_suffix);
+    } catch (std::bad_alloc&) {
+      log_func((char*)"Can not create object string for file %s)", file_name.c_str());
+      return -ENOMEM;
+    }
+    op = std::get<0>(op_data.second);
+    cmpl = std::get<1>(op_data.second);
+    context->aio_operate(obj_name, cmpl, op, 0);
+  }
+
+  for (auto const& op_data: operations) {
+    cmpl = std::get<1>(op_data.second);
+    cmpl->wait_for_complete();
+  }
+  return 0;
+}
+
+ssize_t bulkAioRead::get_results() {
+  /**
+   * Copy the results of executed read requests from ceph's bufferlists to client's buffers
+   *
+   * Note that this method should be called only after the submission and completion of read
+   * requests, i.e. after `submit_and_wait_for_complete` method.
+   *
+   * @return  cumulative number of bytes read (by all read operations) on success, negative
+   *          error code on failure
+   *
+   */
+
+  ceph::bufferlist* bl;
+  char *buf;
+  int *rc;
+  ssize_t res = 0;
+  for (ReadOpData i: buffers) {
+    bl = std::get<0>(i);
+    buf = std::get<1>(i);
+    rc = std::get<2>(i);
+    if (*rc < 0) {
+      //Is it possible to get here?
+      log_func((char*)"One of the reads failed with rc %d", rc);
+      return *rc;
+    }
+    bl->begin().copy(bl->length(), buf);
+    res += bl->length();
+  }
+  return res;
+}
+
+int bulkAioRead::read(void* out_buf, size_t req_size, off64_t offset) {
+  /**
+   * Declare a read operation for file.
+   *
+   * Read coordinates are global, i.e. valid offsets are from 0 to the <file_size> -1, valid request sizes
+   * are from 1 to <file_size> - <offset>. Method can be called multiple times to declare multiple read
+   * operations on the same file.
+   *
+   * @param out_buf    output buffer, where read results should be stored
+   * @param req_size   number of bytes to read
+   * @param offset     offset in bytes where the read should start. Note that the offset is global,
+   *                   i.e. refers to the whole file, not individual ceph objects
+   *
+   * @return  zero on success, negative error code on failure
+   *
+   */
+
+  size_t start_block, last_block, buf_pos, chunk_len, chunk_start, req_len;
+  char *buf_ptr;
+
+  if (req_size == 0) {
+    return -EINVAL;
```

Comment:
return value: zero
Add logging.

Since in principle zero-byte request is valid?